File: public/dsa/searching/search_range/lib/search_range.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> searchRange(vector<int> &arr, int key) {
    // add your logic here
    auto lb = lower_bound(arr.begin(), arr.end(), key);
    auto ub = upper_bound(arr.begin(), arr.end(), key);
    if (lb == arr.end() || *lb != key) {
        return {-1, -1};
    }
    int lbi = lb - arr.begin();
    int ubi = ub - arr.begin();
    return {lbi, ubi - 1};
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 3, 3, 4, 4, 5};
    int key = 3;
    vector<int> indexes = searchRange(arr, key);
    cout << indexes[0] << " " << indexes[1] << endl;
    return 0;
}
// tests-end

/* output
2 4
*/

File: public/dsa/searching/search_range/lib/search_range.py
from bisect import bisect_left, bisect_right
from typing import List


class Solution:
    def searchRange(self, arr: List[int], key: int) -> List[int]:
        index = bisect_left(arr, key)
        l = index if index < len(arr) and arr[index] == key else -1
        if l == -1:
            return [-1, -1]
        r = bisect_right(arr, key) - 1
        return [l, r]


# tests-start
arr = [1, 2, 3, 3, 3, 4, 4, 5]
key = 3
sol = Solution()
print(sol.searchRange(arr, key))
# tests-end

'''output
[2, 4]
'''
